_posts/2020-09-21-porting-namecoin-in-tor-browser-to-macos.md

```diff
@@ -0,0 +1,46 @@
+---
+layout: post
+title: "Porting Namecoin in Tor Browser to macOS"
+author: Jeremy Rand
+tags: [News]
+---
+
+Hot on the heels of [porting Namecoin in Tor Browser to Windows]({{ "/2020/08/28/porting-namecoin-in-tor-browser-to-windows.html" | relative_url }}), it was time to do the final remaining desktop OS, macOS.  This one was always going to present a challenge in the QA testing department, because I don't own any macOS devices.  On the other hand, I fully expected that the amount of code written for macOS support was going to be pretty minimal; in fact I expected that other than tweaking a few constants, macOS support was likely to be a subset of the code needed for Windows support.  But of course, our friend Murphy wasn't going to let that assumption stand.
+
+On the bright side, two of our community members who do have macOS devices volunteered to test my binaries and supply me with debug logs when things failed.  So at least that wasn't going to prevent a successful port.
+
+## Initial Attempt
+
+As a first try, I added the Electrum-NMC, ncprop279, and StemNS dependencies (with their config files) to the `tor-browser` rbm project, and added the missing lines in Tor Launcher for determining the paths of Electrum-NMC and StemNS on macOS.  Then I fliped the `var/namecoin` switch for macOS in `tor-browser-build`, and started up a build job.  In order to build without errors, I then needed to make some slight edits to the build scripts of some of the Go projects that `ncprop279` depends on, mostly related to cgo edge cases.
+
+Alas, the resulting binaries consistently failed to detect Electrum-NMC and ncprop279.
+
+## What is this external user data craziness?
+
+It turns out that Tor Browser for macOS does something very different from Tor Browser for GNU/Linux and Windows.  On macOS, the user data directory is stored separately from the application executable code.  To make matters worse, on macOS there is no way to supply "default" contents for the user data directory.  In fact, in order to set the default bookmarks for Firefox, Tor Browser on macOS does some kind of witchcraft involving stuffing the bookmarks file inside one of the Firefox application archives, such that when Firefox eventually creates a fresh profile, the bookmarks end up in the desired place.  For non-Firefox data, there simply is no mechanism at all.  Even the file `torrc`, which exists in the source code, never actually gets installed on macOS (which I suppose is tolerated because it's an empty file, so who cares if it gets installed?).
```

Comment:
>(which I suppose is tolerated because it's an empty file, so who cares if it gets installed?).
I suppose this is tolerated because it's an empty file; who cares if it gets installed?

Reply:
Doesn't fit my style.